sansa-inference-spark/src/main/scala/net/sansa_stack/inference/spark/forwardchaining/axioms/ForwardRuleReasonerOWLHorst.scala

```diff
@@ -0,0 +1,650 @@
+       .map(a => dataFactory.getOWLClassAssertionAxiom(dataDomainMapBC.value(a.getProperty), a.getSubject))
+       .setName("R4a")
+
+     val R4b = objPropAssertion // .asInstanceOf[RDD[OWLObjectPropertyAssertionAxiom]]
+       .filter(a => objDomainMapBC.value.contains(a.getProperty))
+       .map(a => dataFactory.getOWLClassAssertionAxiom(objDomainMapBC.value(a.getProperty), a.getSubject))
+       .setName("R4b")
+
+     // R5: a rdfs:range x . y a z  -->  z rdf:type x .
+
+     val R5a = dataPropAssertion.filter(a => dataRangeMapBC.value.contains(a.getProperty) && !a.getObject.isLiteral) // Add checking for non-literals
+       .map(a => dataFactory.getOWLClassAssertionAxiom
+     (dataRangeMapBC.value(a.getProperty).asInstanceOf[OWLClassExpression], a.getObject.asInstanceOf[OWLIndividual]))
+       .setName("R5a")
+
+     val R5b = objPropAssertion // .asInstanceOf[RDD[OWLObjectPropertyAssertionAxiom]]
+       .filter(a => objRangeMapBC.value.contains(a.getProperty))
+       .map(a => dataFactory.getOWLClassAssertionAxiom(objRangeMapBC.value(a.getProperty), a.getObject))
+       .setName("R5b")
+
+     // 4. SubClass inheritance according to rdfs9
+     // R6: x rdfs:subClassOf y . z rdf:type x -->  z rdf:type y .
+     val R6 = typeAxioms
+       .filter(a => subClassOfBC.value.contains(a.getClassExpression))
+       .flatMap(a => subClassOfBC.value(a.getClassExpression).map(s => dataFactory.getOWLClassAssertionAxiom(s, a.getIndividual)))
+       .setName("R6")
+
+     // O14: (R owl:hasValue V),(R owl:onProperty P),(X rdf:type R) -> (X P V)
+     // in OWLAxioms we have 3 cases where we can find OWLDataHasValue
+
+     // case a: OWLClassAssertion(OWLDataHasValue(P, w), i)
+     val O14_data_a = typeAxioms.map(a => (a.getClassExpression, a.getIndividual))
+       .filter(a => a._1.isInstanceOf[OWLDataHasValue])
+       .map(a => (a._1.asInstanceOf[OWLDataHasValue], a._2))
+       .map(a => dataFactory.getOWLDataPropertyAssertionAxiom(a._1.getProperty, a._2, a._1.getFiller))
+
+     // case b: OWLSubClassOf(S, OWLDataHasValue(P, w)) , OWLClassAssertion(S, i)
+     val subOperands = subClassof.asInstanceOf[RDD[OWLSubClassOfAxiom]]
+       .map(a => (a.getSubClass, a.getSuperClass))
+
+     val sd = subOperands.filter(s => s._2.isInstanceOf[OWLDataHasValue])
+       .map(s => (s._1, s._2.asInstanceOf[OWLDataHasValue]))
+
+     val O14_data_b = typeAxioms.filter(a => subClassOfBC.value.contains(a.getClassExpression))
+       .map(a => (a.getClassExpression, a.getIndividual))
+       .join(sd).filter(x => x._2._2.isInstanceOf[OWLDataHasValue])
+       .map(a => dataFactory.getOWLDataPropertyAssertionAxiom(a._2._2.getProperty, a._2._1, a._2._2.getFiller))
+
+     // case c: OWLEquivelantClass(E, OWLDataHasValue(P, w)), OWLClassAssertion(E, i)
+     val eqOperands = equClass.asInstanceOf[RDD[OWLEquivalentClassesAxiom]]
+       .map(a => (a.getOperandsAsList.get(0), a.getOperandsAsList.get(1)))
```

Comment:
There are a few of such cases where you assume there will be exactly two operands. However, there can be many. You will first have to get all equivalent classes/properties/... as pairwise axioms via `.flatMap(equivClassesAxiom => equivClassesAxiom.asPairwiseAxioms().asScala)` as exemplified above in rule 011a/012b.

Reply:
For my knowledge, I think it is only 2 equivalent class. Thanks for clarification